be/queue.c

```diff
@@ -0,0 +1,448 @@
+	M0_ASSERT_INFO(bq->bq_enqueued >= bq->bq_dequeued,
+	               "bq="BEQ_F, BEQ_P(bq));
+	return bq->bq_enqueued - bq->bq_dequeued;
+}
+
+static bool be_queue_is_empty(struct m0_be_queue *bq)
+{
+	return be_queue_q_size(bq) == 0;
+}
+
+static bool be_queue_is_full(struct m0_be_queue *bq)
+{
+	return be_queue_q_size(bq) >= bq->bq_cfg.bqc_q_size_max;
+}
+
+static struct be_queue_item *be_queue_q_put(struct m0_be_queue  *bq,
+                                            const struct m0_buf *data)
+{
+	struct be_queue_item *bqi;
+
+	M0_PRE(m0_mutex_is_locked(&bq->bq_lock));
+
+	bqi = bqq_tlist_head(&bq->bq_q_unused);
+	m0_buf_memcpy(&BE_QUEUE_ITEM2BUF(bq, bqi), data);
+	bqq_tlist_move_tail(&bq->bq_q, bqi);
+	++bq->bq_enqueued;
+	M0_LEAVE("bq="BEQ_F, BEQ_P(bq));
+	return bqi;
+}
+
+static void be_queue_q_peek(struct m0_be_queue *bq, struct m0_buf *data)
+{
+	struct be_queue_item *bqi;
+
+	M0_PRE(m0_mutex_is_locked(&bq->bq_lock));
+	M0_PRE(!be_queue_is_empty(bq));
+
+	bqi = bqq_tlist_head(&bq->bq_q);
+	m0_buf_memcpy(data, &BE_QUEUE_ITEM2BUF(bq, bqi));
+	M0_LEAVE("bq="BEQ_F, BEQ_P(bq));
+}
+
+static void be_queue_q_get(struct m0_be_queue *bq,
+			   struct m0_buf      *data,
+                           bool               *successful)
+{
+	struct be_queue_item *bqi;
+
+	M0_PRE(m0_mutex_is_locked(&bq->bq_lock));
+	M0_PRE(!be_queue_is_empty(bq));
+
+	bqi = bqq_tlist_head(&bq->bq_q);
+	m0_buf_memcpy(data, &BE_QUEUE_ITEM2BUF(bq, bqi));
+	*successful = true;
+	bqq_tlist_move(&bq->bq_q_unused, bqi);
+	++bq->bq_dequeued;
+	M0_LEAVE("bq="BEQ_F, BEQ_P(bq));
+}
+
+static void be_queue_op_put(struct m0_be_queue   *bq,
+                            struct m0_be_op      *op,
+                            struct be_queue_item *bqi)
+{
+	struct be_queue_wait_op *bwo;
+
+	M0_PRE(m0_mutex_is_locked(&bq->bq_lock));
+	M0_PRE(!bqop_tlist_is_empty(&bq->bq_op_put_unused));
+
+	bwo = bqop_tlist_head(&bq->bq_op_put_unused);
+	bwo->bbo_bqi = bqi;
+	bwo->bbo_op  = op;
+	bqop_tlist_move_tail(&bq->bq_op_put, bwo);
+	M0_LEAVE("bq="BEQ_F, BEQ_P(bq));
+}
+
+static void be_queue_op_put_done(struct m0_be_queue *bq)
+{
+	struct be_queue_wait_op *bwo;
+
+	M0_PRE(m0_mutex_is_locked(&bq->bq_lock));
+	M0_PRE(!bqop_tlist_is_empty(&bq->bq_op_put));
+
+	bwo = bqop_tlist_head(&bq->bq_op_put);
+	m0_be_op_done(bwo->bbo_op);
+	bqop_tlist_move(&bq->bq_op_put_unused, bwo);
+	M0_LEAVE("bq="BEQ_F, BEQ_P(bq));
+}
+
+static bool be_queue_op_put_is_waiting(struct m0_be_queue *bq)
+{
+	return !bqop_tlist_is_empty(&bq->bq_op_put);
+}
+
+static void be_queue_op_get(struct m0_be_queue *bq,
+                            struct m0_be_op    *op,
+                            struct m0_buf      *data,
+                            bool               *successful)
+{
+	struct be_queue_wait_op *bwo;
+
+	M0_PRE(m0_mutex_is_locked(&bq->bq_lock));
+	M0_PRE(!bqop_tlist_is_empty(&bq->bq_op_get_unused));
+
+	bwo = bqop_tlist_head(&bq->bq_op_get_unused);
+	bwo->bbo_data       = *data;
```

Comment:
I propose to add:
M0_ASSERT_INFO(bwo != NULL, "Too many consumers");

Reply:
Done.